domain/application/service/unit.go

```diff
@@ -226,6 +231,32 @@ func (s *Service) UpdateCAASUnit(ctx context.Context, unitName coreunit.Name, pa
 	return nil
 }
 
+// SetReportedUnitAgentVersion sets the reported agent version for the
+// supplied unit name. Reported agent version is the version that the agent
+// binary on this unit has reported it is running.
+//
+// The following errors are possible:
+// - [coreerrors.NotValid] - when the reportedVersion is not valid.
+// - [coreerrors.NotSupported] - when the architecture is not supported.
+// - [applicationerrors.UnitNotFound] - when the unit does not exist.
+// - [applicationerrors.UnitIsDead] - when the unit is dead.
+func (s *Service) SetReportedUnitAgentVersion(ctx context.Context, unitUUID coreunit.UUID, reportedVersion agentbinary.Version) error {
+	if err := reportedVersion.Validate(); err != nil {
+		return errors.Errorf("reported agent version %v is not valid: %w", reportedVersion, err)
+	}
+
+	if err := s.st.SetRunningAgentBinaryVersion(ctx, unitUUID, reportedVersion); err != nil {
+		return errors.Errorf(
+			"setting unit %q reported agent version (%s) in state: %w",
+			unitUUID.String(),
+			reportedVersion.Number.String(),
```

Comment:
Because we are trying to provide context here and be helpful maybe `agentbinary.Version` should implant the stringer interface so it can not only output the version number but also the arch so we don't drop important information here.